### PR TITLE
Add NearestNeighbors SPMD API

### PR DESCRIPTION
**Pull request by ethanglaser**
## Description

Adds spmd NearestNeighbors to API with necessary modifications to sklearnex/spmd and onedal/spmd, along with minor revisions in onedal4py and onedal itself (https://github.com/uxlfoundation/oneDAL/pull/3262 - which is prerequisite for merging this). Test also added for validation.

Full list of changes:
- Support raw inputs for kneighbors function
- Remove weights from NearestNeighbors class (sklearn does not support this nor does it logically make sense)
- Add NearestNeighbors to API for sklearnex and onedal spmd modules
- Enable spmd usage of kneighbors in all knn classes (added storage of queue from fit to use if X is None in kneighbors())
- Revert incorrect usage of _assert_unordered_allclose for _spmd_assert_allclose in neighbors comparisons in spmd test
- Add gold and synthetic tests for spmd NearestNeighbors, including large test that revealed sycl event issue in oneDAL that has been addressed in https://github.com/uxlfoundation/oneDAL/pull/3262
- Added test scope for kneighbors with X=None (this would have failed previously)

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by ethanglaser** [comment]
Combined CI: http://intel-ci.intel.com/f04c5cd4-6c88-f193-9c31-a4bf010d0e2d

File: sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py [hunk 1]
```diff
@@ -308,3 +309,83 @@ def test_knnreg_spmd_synthetic(
...
+    get_dataframes_and_queues(dataframe_filter_="dpnp,dpctl", device_filter_="gpu"),
+)
+@pytest.mark.mpi
+def test_knnsearch_spmd_gold(dataframe, queue):
+    # Import spmd and batch algo
+    from sklearnex.neighbors import NearestNeighbors as NearestNeighbors_Batch
+    from sklearnex.spmd.neighbors import NearestNeighbors as NearestNeighbors_SPMD
+
+    # Create gold data and convert to dataframe
+    X_train = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
```
Comment (ethanglaser) [inline sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py]:
```suggestion
    X_train = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2], [10, 10], [9, 9]])
```

File: sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py [hunk 2]
```diff
@@ -308,3 +309,83 @@ def test_knnreg_spmd_synthetic(
...
+    get_dataframes_and_queues(dataframe_filter_="dpnp,dpctl", device_filter_="gpu"),
+)
+@pytest.mark.mpi
+def test_knnsearch_spmd_gold(dataframe, queue):
+    # Import spmd and batch algo
+    from sklearnex.neighbors import NearestNeighbors as NearestNeighbors_Batch
+    from sklearnex.spmd.neighbors import NearestNeighbors as NearestNeighbors_SPMD
+
+    # Create gold data and convert to dataframe
+    X_train = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2], [10, 10], [9, 9]])
```
Comment (ethanglaser) [inline sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py]:
```suggestion
    X_train = np.array(
        [[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2], [10, 10], [9, 9]]
    )
```
```suggestion
    X_train = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2], [10, 10], [9, 9]])
```

File: sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py [hunk 3]
```diff
@@ -308,3 +309,85 @@ def test_knnreg_spmd_synthetic(
...
+    _assert_unordered_allclose(spmd_indcs, batch_indcs, localize=True)
+    _assert_unordered_allclose(spmd_dists, batch_dists, localize=True)
+
+
+@pytest.mark.skipif(
+    not _mpi_libs_and_gpu_available,
+    reason="GPU device and MPI libs required for test",
+)
+@pytest.mark.parametrize(
+    "dimensions", [{"n": 100, "m": 10, "k": 2}, {"n": 100000, "m": 100, "k": 100}]
```
Comment (ethanglaser) [inline sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py:348]:
```suggestion
    "dimensions", [{"n": 100, "m": 10, "k": 2}, {"n": 20000, "m": 100, "k": 100}]
```

**Comment by ethanglaser** [comment]
CI (with https://github.com/uxlfoundation/oneDAL/pull/3262): http://intel-ci.intel.com/f09afe5b-dc4b-f1cf-81b6-d4f5ef20c6a0

File: sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py [hunk 4]
```diff
@@ -303,8 +301,133 @@ def test_knnreg_spmd_synthetic(
...
+    local_dpt_y_train = _convert_to_dataframe(
+        _get_local_tensor(y_train), sycl_queue=queue, target_df=dataframe
+    )
+
+    # Run each estimator without an input to kneighbors() and ensure functionality and equivalence
+    for CurrentEstimator in [KNeighborsClassifier, KNeighborsRegressor, NearestNeighbors]:
+        spmd_model = CurrentEstimator(n_neighbors=2, algorithm="brute").fit(
+            local_dpt_X_train, local_dpt_y_train
+        )
+        batch_model = NearestNeighbors_Batch(n_neighbors=2, algorithm="brute").fit(
```
Comment (ethanglaser) [inline sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py]:
```suggestion
        spmd_model = CurrentEstimator(n_neighbors=1, algorithm="brute").fit(
            local_dpt_X_train, local_dpt_y_train
        )
        batch_model = NearestNeighbors_Batch(n_neighbors=1, algorithm="brute").fit(
```

File: onedal/neighbors/neighbors.py [hunk 5]
```diff
@@ -730,7 +732,6 @@ def __init__(
         self,
         n_neighbors=5,
         *,
-        weights="uniform",
```
Comment (icfaust) [inline onedal/neighbors/neighbors.py:733]: Info on this change would be nice (should be added to the description of the PR)

File: onedal/neighbors/neighbors.py [hunk 6]
```diff
@@ -792,7 +793,7 @@ def _onedal_predict(self, model, X, params):
         return self.infer(params, model, X)
 
     @supports_queue
-    def fit(self, X, y, queue=None):
+    def fit(self, X, y=None, queue=None):
```
Comment (icfaust) [inline onedal/neighbors/neighbors.py:796]: I assume previously the ```y``` value was only set to ```None``` from only sklearnex calls (otherwise this should fail in some way in ```onedal``` testing. It looks like there is no independent testing of the kNeighbors algos beyond classification in the ```onedal``` module. This doesn't need to be done, but something we should attempt to maybe address in #2700

File: onedal/spmd/neighbors/neighbors.py [hunk 7]
```diff
@@ -30,6 +31,7 @@ def infer(self, *args, **kwargs): ...
 
     @support_input_format
     def fit(self, X, y, queue=None):
+        self.spmd_queue_ = queue
```
Comment (icfaust) [inline onedal/spmd/neighbors/neighbors.py:35]: Should add a code comment here why this exists (to save future developers time).

File: sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py [hunk 8]
```diff
@@ -303,8 +301,133 @@ def test_knnreg_spmd_synthetic(
...
+    y_train = np.array([0, 1, 0, 1, 0, 1, 0, 1])
+    local_dpt_X_train = _convert_to_dataframe(
+        _get_local_tensor(X_train), sycl_queue=queue, target_df=dataframe
+    )
+    local_dpt_y_train = _convert_to_dataframe(
+        _get_local_tensor(y_train), sycl_queue=queue, target_df=dataframe
+    )
+
+    # Run each estimator without an input to kneighbors() and ensure functionality and equivalence
+    for CurrentEstimator in [KNeighborsClassifier, KNeighborsRegressor, NearestNeighbors]:
```
Comment (icfaust) [inline sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py:422]: I see why this was done, but is a bit painful to analyze if there is a failure. Ideally it would be parametrized over, but really isn't possible by the way it is imported. Would be worth adding some sort of message to figure out which is the CurrentEstimator (rather than having to dig through the pytest log for the CurrentEstimator current value was).

File: sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py [hunk 9]
```diff
@@ -303,8 +301,133 @@ def test_knnreg_spmd_synthetic(
...
+    spmd_model = NearestNeighbors_SPMD(
+        n_neighbors=dimensions["k"], algorithm="brute"
+    ).fit(local_dpt_X_train)
+    batch_model = NearestNeighbors_Batch(
+        n_neighbors=dimensions["k"], algorithm="brute"
+    ).fit(X_train)
+    spmd_dists, spmd_indcs = spmd_model.kneighbors(local_dpt_X_train)
+    batch_dists, batch_indcs = batch_model.kneighbors(X_train)
+
+    tol = 0.005 if dtype == np.float32 else 1e-6
```
Comment (icfaust) [inline sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py:386]: Yikes on this float32 setting. Any info on it? Especially because there is a skip associated with it above (meaning an even worse value occurs?)

**Review by icfaust — APPROVED** [review]
Due to time constraints I have tried to push most of my changes into follow-on PRs. There are some nagging issues we should talk through, but I see no reason to hold off getting this in.

File: onedal/neighbors/neighbors.py (hunk 5 above)
Comment (ethanglaser) [reply onedal/neighbors/neighbors.py:733]: This would've been mistakenly added (likely because cls and reg have weights param) as sklearn's NearestNeighbors does not have weights parameter

File: onedal/neighbors/neighbors.py (hunk 6 above)
Comment (ethanglaser) [reply onedal/neighbors/neighbors.py:796]: Looks like test scope is really lacking here. I can create a ticket.

File: onedal/spmd/neighbors/neighbors.py (hunk 7 above)
Comment (ethanglaser) [reply onedal/spmd/neighbors/neighbors.py:35]: good point, added

File: sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py (hunk 9 above)
Comment (ethanglaser) [reply sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py:386]: It's true, and good observation. It's pretty tricky because this assert all close functionality will fail even if a single element is not within the threshold, hence why it is so loose - it would be nice if there was some sort of customization of that.

It's possible that we could still run the indices check for this case, but distances are more fragile.

Comment (ethanglaser) [reply sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py:386]: This is not the only place in spmd test scope where drastically low thresholds are needed to support float32 tests passing though

File: sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py (hunk 8 above)
Comment (ethanglaser) [reply sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py:422]: Yeah - I am pretty open to ideas on this one. The loop is great because I run the exact same test on all 3 classes, but you are correct that analysis on a fail is trickier. I think scikit-learn may do things like this, I could check how they do it.

Comment (ethanglaser) [reply sklearnex/spmd/neighbors/tests/test_neighbors_spmd.py:422]: I guess its easier there because in sklearn they import at top of file

**Comment by ethanglaser** [comment]
Final CI: http://intel-ci.intel.com/f09e0f06-8f7b-f1e2-8c25-a4bf010d0e2d
